hack/kind-up.sh

```diff
@@ -167,6 +172,88 @@ check_registry_cache_availability() {
+# so that we can later scrape the various `kubelet_volume_stats_*' metrics
+# exposed by kubelet (see [5]).
+#
+# References:
+#
+# [1]: https://github.com/rancher/local-path-provisioner
+# [2]: https://kubernetes.io/docs/concepts/storage/volumes/#hostpath
+# [3]: https://kubernetes.io/docs/concepts/storage/volumes/#local
+# [4]: https://github.com/kubernetes-sigs/kind/blob/main/pkg/cluster/internal/create/actions/installstorage/storage.go
+# [5]: https://kubernetes.io/docs/reference/instrumentation/metrics/
+_setup_kind_sc_default_volume_type() {
+    echo "Configuring default StorageClass for kind cluster ..."
+    kubectl get storageclasses standard -o yaml | \
+        yq '.metadata.annotations.defaultVolumeType = "local"' | \
+        kubectl apply -f -
+}
+
+# The rancher.io/local-path provisioner at the moment does not support volume
+# resizing (see [1]). There is an open PR, which is scheduled for the next
+# release around May, 2024 (see [2]). Until [2] is merged we will use a custom
+# local-path provisioner with support for volume resizing.
+#
+# This function should be called after setting up the containerd registries on
+# the kind nodes.
+#
+# References:
+#
+# [1]: https://github.com/rancher/local-path-provisioner
+# [2]: https://github.com/rancher/local-path-provisioner/pull/350
+_setup_kind_with_lpp_resize_support() {
+    if [ "${WITH_LPP_RESIZE_SUPPORT}" != "true" ]; then
+        return
+    fi
+
+    echo "Configuring kind local-path provisioner with volume resize support ..."
+
+    # First configure allowVolumeExpansion on the default StorageClass
+    kubectl get storageclasses standard -o yaml | \
+        yq '.allowVolumeExpansion = true' | \
+        kubectl apply -f -
+
+    local _workdir=$( mktemp -d )
+    local _oldpwd="${PWD}"
+
+    # The fork of LPP with resize support
+    local _lpp_repo="https://github.com/marjus45/local-path-provisioner.git"
+    local _lpp_branch="feature-external-resizer"
+
+    # Build the latest image
+    cd "${_workdir}" && \
+        git clone "${_lpp_repo}" && \
+        cd local-path-provisioner && \
+        git checkout -b "${_lpp_branch}" "origin/${_lpp_branch}" && \
+        make
+
+    # Push to the local registry
+    local _latest_image=$( cat ./bin/latest_image )
+    local _local_latest_image="localhost:5001/${_latest_image}"
+    docker image tag "${_latest_image}" "${_local_latest_image}"
+    docker push "${_local_latest_image}"
+
+    # Apply the latest manifests and use our own image
+    cd deploy && \
+        kustomize edit set image rancher/local-path-provisioner:master-head="${_local_latest_image}" && \
+        kustomize build . | \
+        kubectl apply -f -
+
+    # The default manifests from rancher/local-path come with another
+    # StorageClass, which we don't need, so make sure to remove it.
+    kubectl delete --ignore-not-found=true storageclass local-path
+
+    cd "${_oldpwd}"
+    rm -rf "${_workdir}"
```

Comment:
Can you elaborate why we need this approach to build the local-path-provisioner from source?

For example, in https://github.com/gardener/gardener/pull/9048 we had a similar case. @oliver-goetz prepared a container image of the upstream distribution/distribution project (`ghcr.io/oliver-goetz/distribution/registry:3.0.0-dev`) and adapted this image to be used instead until there is an upstream release that contains his fix.
Can't we follow the same pattern here? I mean to build a our custom image from the above PR and patch the local-path-provisioner Deployment to use that image with `kubectl patch`.

Reply:
Going with a custom image was one of the options I've also considered, but decided to drop it :)

The reason why I didn't go with a custom image is because, it's not only the image itself that we need, but also additional kustomization manifests, e.g. `ClusterRole`, `ConfigMap`, etc. 

We could embed these into the new custom image as well, but then we need to take them out and apply them anyways. Another approach would be to pull these new `ClusterRole`, `ConfigMap` (and other manifests), into `gardener/gardener`, but that again would mean to pull in additional dependencies, which we need to drop soon, so I've decided it's better not to pull in more additional dependencies and complicate things needlessly.

Let me know if still think that going with a custom image + the rest of the manifests is something that is worth going for, considering that we may end up complicating things even more.

Reply:
I was looking into the topic and I ended up with https://github.com/ialidzhikov/gardener/commit/ca5b5cc2380a7e8bd45892a62849560728224e4c. With the pre-built container image we don't need to clone the repo, build the local-path-provisioner images, push them etc. It should also speed up the `DEV_SETUP_WITH_LPP_RESIZE_SUPPORT=true make kind-up` execution time and in general involve one idea less hacks during `make kind-up`. WDYT?

I used the Dockerfile in https://github.com/kubernetes-sigs/kind/tree/main/images/local-path-provisioner to build a multi-arch local-path-provisioner image.
I will now try to push the image to GHCR as according to the component checklist we don't use dockerhub images.

Reply:
> I will now try to push the image to GHCR as according to the component checklist we don't use dockerhub images.

The ghcr image is `ghcr.io/ialidzhikov/local-path-provisioner:feature-external-resizer-c0c1c13`.

Reply:
I'm more in favour of @ialidzhikov's suggestion. You could just add the `-sSfL` options to the curl command.

Reply:
> I was looking into the topic and I ended up with [ialidzhikov@ca5b5cc](https://github.com/ialidzhikov/gardener/commit/ca5b5cc2380a7e8bd45892a62849560728224e4c). With the pre-built container image we don't need to clone the repo, build the local-path-provisioner images, push them etc. It should also speed up the `DEV_SETUP_WITH_LPP_RESIZE_SUPPORT=true make kind-up` execution time and in general involve one idea less hacks during `make kind-up`. WDYT?
> 

I'm okay with this approach.

> PS: We also don't need kustomize with the above suggestion.

I think we should still use it. The `deploy/local-path-storage.yaml` is part of a `kustomization` bundle, so any changes that happen in the meantime until the PR is merged into upstream may end up breaking the local setup, e.g. adding new manifests to `kustomization.yaml` or splitting the `local-path-storage.yaml`, etc. will not be captured and applied.